### PR TITLE
Handle secret/keyed hashing

### PR DESCRIPTION
**Pull request by Gadzy**
Provides secret/key hashing support with direct binding to the underlying argon2 implementation.

See (https://github.com/ranisalt/node-argon2/issues/58) and (https://github.com/ranisalt/node-argon2/pull/137)

### CODE REVIEW

**Comment by ranisalt** [comment]
Sweet! How did you generate the hashes, though?

**Comment by Gadzy** [comment]
I used the following c code using the same implementation as the repo.

Although I think using the KATs provided [here](https://github.com/P-H-C/phc-winner-argon2/tree/master/kats) could be better. I could submit a PR if you like.

```c
#include "argon2.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define HASHLEN 32
#define SALTLEN 16
#define PWD "password"

int main(void)
{
    uint8_t hash1[HASHLEN];
    uint8_t hash2[HASHLEN];

    uint8_t *salt = (uint8_t *)strdup("saltsaltsaltsalt");

    uint8_t *secret = (uint8_t *)strdup("secretsecretsecr");

    uint8_t *pwd = (uint8_t *)strdup(PWD);
    uint32_t pwdlen = strlen((char *)pwd);

    uint32_t t_cost = 3;            // 1-pass computation
    uint32_t m_cost = (1<<12);      // 64 mebibytes memory usage
    uint32_t parallelism = 1;       // number of threads and lanes

    // low-level API
    argon2_context context = {
        hash,  /* output array, at least HASHLEN in size */
        HASHLEN, /* digest length */
        pwd, /* password array */
        pwdlen, /* password length */
        salt,  /* salt array */
        SALTLEN, /* salt length */
        secret, 16, /* optional secret data */
        NULL, 0, /* optional associated data */
        t_cost, m_cost, parallelism, parallelism,
        ARGON2_VERSION_13, /* algorithm version */
        NULL, NULL, /* custom memory allocation / deallocation functions */
        /* by default only internal memory is cleared (pwd is not wiped) */
        ARGON2_DEFAULT_FLAGS
    };

    int rc = argon2i_ctx( &context );
    if(ARGON2_OK != rc) {
        printf("Error: %s\n", argon2_error_message(rc));
        exit(1);
    }
    free(pwd);

    for( int i=0; i<HASHLEN; ++i ) printf( "%02x", hash[i] ); printf( "\n" );
    return 0;
}
```

**Comment by ranisalt** [comment]
I guess using the kats would be better but I have no clue how that file works.

File: test/test.js [hunk 1]
```diff
@@ -77,6 +79,16 @@ describe('Argon2', () => {
       const hash = await argon2.hash(password, { associatedData, salt })
       assert.equal(hashes.withAd, hash)
     })
+
+    it('with associated data', async () => {
+      const hash = await argon2.hash(password, { associatedData, salt })
+      assert.equal(hashes.withAd, hash)
+    })
+
```
Comment (ranisalt) [inline test/test.js]: Great resolution algorith, git!

This duped test needs to be removed...

Comment (Gadzy) [reply test/test.js]: > I guess using the kats would be better but I have no clue how that file works.

Working on it ATM

> Great resolution algorith, git!
> 
> This duped test needs to be removed...

Haha 😄 Will do !

**Comment by ranisalt** [comment]
Thanks!

**Pull request merged**